utils/src/chunked_polynomial.rs

```diff
@@ -0,0 +1,164 @@
+use ark_ff::{Field, Zero};
+use ark_poly::polynomial::{univariate::DensePolynomial, Polynomial, UVPolynomial};
+
+use rayon::prelude::*;
+
+#[derive(Clone)]
+pub struct ChunkedPolynomials<P> {
+    pub polys: Vec<P>,
+    pub degree: usize,
+}
+
+impl<P> Default for ChunkedPolynomials<P> {
+    fn default() -> ChunkedPolynomials<P> {
+        ChunkedPolynomials {
+            polys: vec![],
+            degree: 0,
+        }
+    }
+}
+
+impl<P> ChunkedPolynomials<P> {
+    pub fn add_chunk(&mut self, p: P) {
+        self.polys.push(p)
+    }
+}
+
+impl<F: Field> ChunkedPolynomials<DensePolynomial<F>> {
+    /// This function evaluates polynomial in chunks.
+    pub fn linearize(&self, elm: F) -> Vec<F> {
+        let mut res: Vec<F> = vec![];
+        for poly in self.polys.iter() {
+            let eval = poly.evaluate(&elm);
+            res.push(eval);
+        }
+        res
+    }
+
+    /// Multiplies the chunks of a polynomial with powers of zeta^n to make it of degree n-1.
+    /// For example, if a polynomial can be written `f = f0 + x^n f1 + x^2n f2`
+    /// (where f0, f1, f2 are of degree n-1), then this function returns the new semi-evaluated
+    /// `f'(x) = f0(x) + zeta^n f1(x) + zeta^2n f2(x)`.
+    pub fn compress_polynomial(&self, zeta_n: F) -> DensePolynomial<F> {
+        let mut scale = F::one();
+        let mut coeffs = vec![F::zero(); self.degree];
+
+        for poly in self.polys.iter() {
+            for (coeff, poly_coeff) in coeffs.iter_mut().zip(&poly.coeffs) {
+                *coeff += scale * poly_coeff;
+            }
+
+            scale *= zeta_n;
+        }
+
+        while coeffs.last().map_or(false, |c| c.is_zero()) {
+            coeffs.pop();
+        }
```

Comment:
> I feel like this could be rewritten in a clearer way. Perhaps something like this:
> 
> ```rust
> let coeffs = coeffs.rev().skip_while(|c| c.is_zero()).rev().collect();
> ```

well, I copy the codes here... Yes, the new one is better i think.

Reply:
> I feel like this could be rewritten in a clearer way. Perhaps something like this:
> 
> ```rust
> let coeffs = coeffs.rev().skip_while(|c| c.is_zero()).rev().collect();
> ```

This is not the functionality we want. Here we remove the leading zeros, while the code above remove all the zero coefficient.
Actually, I just copy the codes in the original implementation. :)

Reply:
I think they're all equivalent :o https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e7dda821e6964961bdb6d350929edaac

> Actually, I just copy the codes in the original implementation. :)

yeah sorry about that ^^